constraints_lang.md

```diff
@@ -0,0 +1,311 @@
+# Motivation
+
+When users of CAD software construct models, they do it in whatever order is quickest. They do not generally draw connected segments or add constraints in any particular order.
+
+Also, when thinking geometrically, users typically think about relationships between geometry. For example, these two lines are parallel. This line and this arc are tangent. They _don't_ typically think in terms of functions that compute the locations of points given other points.
+
+In contrast, KCL of today has strict requirements of the order of things and how dependencies are defined. The equal length constraint in the Modeling App makes it so that the length of a segment is derived from the length of another segment. We require that users write expressions computing the derived value given the depended upon values. This means that paths must be constructed in a specific order that's a topological sort of dependencies. In particular, the first segment, which we'll call the root, cannot depend on any subsequent segments.
+
+Additionally, KCL requires users to write functions that ultimately come down to computing the locations of points given other points. We have standard library functions that help make this somewhat higher-level, like `tangentialArc`, but it ultimately comes down to directly computing output locations based on inputs. This means that users must adapt their way of thinking to the tool rather than the tool adapting to them.
+
+Notation: ||AB|| is defined as the length of the line segment between points A and B.
```

Comment:
I think the `||v1||` notation would be opaque to most users. It's clear for your proposal but I think having the default notation be `len(v1)` or something would be more immediately understandable.

Reply:
I'd argue that this is one of the few notations from math that's standard enough to use. Pretty sure we learned this in high school. There was talk of using `||` for logical OR. It would be nice if we got away from general purpose programming language syntax and leaned-in to geometry and linear algebra notation.

But I don't care for this notation in particular if people don't like it, just the general idea. There are infinite possibilities, and everyone tends to use the same old thing. So I will continue to propose strange-at-first-glance language ideas 🙂

Reply:
Totally! I like that as a general rule, lean into the syntax of geometry. If our mechies understand it at a glance I'd be all for it. I think I've been poisoned by seeing logical ORs as `||` for too long, so that magnitude notation has become strange to me since high school, but you're right.